src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -5669,8 +5685,24 @@ and TcExprUndelayed (cenv: cenv) (overallTy: OverallTy) env tpenv (synExpr: SynE
         TcExprObjectExpr cenv overallTy env tpenv (synObjTy, argopt, binds, extraImpls, mNewExpr, m)
 
     | SynExpr.Record (inherits, withExprOpt, synRecdFields, mWholeExpr) ->
-        TcNonControlFlowExpr env <| fun env ->
-        TcExprRecord cenv overallTy env tpenv (inherits, withExprOpt, synRecdFields, mWholeExpr)
+        // When the original expression in copy-and-update is more complex than `{ x with ... }`, like `{ f () with ... }`,
+        // bind it first, so that it's not evaluated multiple times during a nested update
+        match withExprOpt with
+        | None
+        | Some(SynExpr.Ident _, _) ->
+            TcNonControlFlowExpr env <| fun env ->
+            TcExprRecord cenv overallTy env tpenv (inherits, withExprOpt, synRecdFields, mWholeExpr)
+        | Some(expr, blockSep) ->
+            let mOrigExprSynth = expr.Range.MakeSynthetic()
+            let id = mkSynId mOrigExprSynth "bind@"
+            let binding = mkSynBinding (PreXmlDoc.Empty, mkSynPatVar None id) (None, false, false, mOrigExprSynth, DebugPointAtBinding.NoneAtSticky, None, expr, mOrigExprSynth, [], [], None, SynBindingTrivia.Zero)
+
+            let withExpr = SynExpr.Ident id, blockSep
+
+            let body = SynExpr.Record (inherits, Some withExpr, synRecdFields, mWholeExpr)
+            let expr = SynExpr.LetOrUse (false, false, [ binding ], body, mOrigExprSynth, SynExprLetOrUseTrivia.Zero)
```

Comment:
Can the shared body of the repeated block between AnonRecord and Record be extracted please?